File: src/android/AzureUpload.java
package com.karamsawalha.azureupload;

import android.content.Context;
import android.graphics.Bitmap;
import android.media.MediaMetadataRetriever;
import android.util.Base64;
import android.util.Log;
import android.widget.Toast;
import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.PluginResult;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class AzureUpload extends CordovaPlugin {

    private static final String STORAGE_URL = "https://arabicschool.blob.core.windows.net/";
    private static final String TAG = "AzureUpload";

    @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
        if (action.equals("uploadFiles")) {
            String postId = args.getString(0);
            String sasToken = args.getString(1);
            JSONArray files = args.getJSONArray(2);
            uploadFiles(postId, sasToken, files, callbackContext);
            return true;
        }
        return false;
    }

    private void uploadFiles(String postId, String sasToken, JSONArray files, CallbackContext callbackContext) {
        ExecutorService executorService = Executors.newFixedThreadPool(files.length());

        for (int i = 0; i < files.length(); i++) {
            try {
                JSONObject file = files.getJSONObject(i);
                String fileName = file.getString("filename");
                String originalName = file.getString("originalname");
                String mimeType = file.getString("mimetype");
                String binaryData = file.getString("binarydata");
                String thumbnailName = file.has("thumbnail") ? file.getString("thumbnail") : "";

                if (mimeType.startsWith("image")) {
                    executorService.execute(() -> uploadImage(fileName, binaryData, sasToken, originalName, postId));
                } else if (mimeType.startsWith("video")) {
                    executorService.execute(() -> uploadVideo(fileName, binaryData, thumbnailName, sasToken, originalName, postId));
                } else {
                    executorService.execute(() -> uploadFile(fileName, binaryData, sasToken, originalName, postId));
                }
            } catch (JSONException e) {
                Log.e(TAG, "Error processing file upload: " + e.getMessage());
            }
        }

        executorService.shutdown();
        callbackContext.success("File uploads initiated.");
    }

    private void uploadImage(String fileName, String binaryData, String sasToken, String originalName, String postId) {
        try {
            byte[] imageData = Base64.decode(binaryData, Base64.DEFAULT);
            uploadToAzure(fileName, imageData, sasToken, originalName, postId);
        } catch (Exception e) {
            Log.e(TAG, "Error uploading image: " + e.getMessage());
        }
    }

    private void uploadVideo(String fileName, String binaryData, String thumbnailName, String sasToken, String originalName, String postId) {
        try {
            // Decode base64 video data
            byte[] videoData = Base64.decode(binaryData, Base64.DEFAULT);
            
            // Write the video data to a temporary file
            File tempVideoFile = File.createTempFile("video", ".mp4", cordova.getContext().getCacheDir());
            FileOutputStream fos = new FileOutputStream(tempVideoFile);
            fos.write(videoData);
            fos.close();

            // Upload the video to Azure
            uploadToAzure(fileName, videoData, sasToken, originalName, postId);

            // Use MediaMetadataRetriever to generate a thumbnail from the temporary file
            MediaMetadataRetriever retriever = new MediaMetadataRetriever();
            retriever.setDataSource(tempVideoFile.getAbsolutePath());

            Bitmap bitmap = retriever.getFrameAtTime(0);
            ByteArrayOutputStream out = new ByteArrayOutputStream();
            bitmap.compress(Bitmap.CompressFormat.PNG, 100, out);
            byte[] thumbnailData = out.toByteArray();

            // Upload the thumbnail to Azure
            uploadToAzure(thumbnailName, thumbnailData, sasToken, originalName, postId);

            // Delete the temporary video file after use
            tempVideoFile.delete();
        } catch (Exception e) {
            Log.e(TAG, "Error uploading video: " + e.getMessage());
        }
    }

    private void uploadFile(String fileName, String binaryData, String sasToken, String originalName, String postId) {
        try {
            byte[] fileData = Base64.decode(binaryData, Base64.DEFAULT);
            uploadToAzure(fileName, fileData, sasToken, originalName, postId);
        } catch (Exception e) {
            Log.e(TAG, "Error uploading file: " + e.getMessage());
        }
    }

    private void uploadToAzure(String fileName, byte[] fileData, String sasToken, String originalName, String postId) {
        try {
            String urlString = STORAGE_URL + fileName + "?" + sasToken;
            URL url = new URL(urlString);
            HttpURLConnection connection = (HttpURLConnection) url.openConnection();
            connection.setDoOutput(true);
            connection.setRequestMethod("PUT");
            connection.setRequestProperty("Content-Type", "application/octet-stream");

            connection.getOutputStream().write(fileData);

            int responseCode = connection.getResponseCode();
            if (responseCode == HttpURLConnection.HTTP_CREATED || responseCode == HttpURLConnection.HTTP_OK) {
                commitUpload(postId, STORAGE_URL + fileName, originalName, "application/octet-stream");
            } else {
                Log.e(TAG, "Error uploading to Azure, response code: " + responseCode);
            }
        } catch (Exception e) {
            Log.e(TAG, "Error uploading to Azure: " + e.getMessage());
        }
    }

    private void commitUpload(String postId, String fileUrl, String originalName, String mimeType) {
        try {
            URL commitUrl = new URL("https://personal-fjlz3d21.outsystemscloud.com/uploads/rest/a/commit?postid=" + postId);
            HttpURLConnection connection = (HttpURLConnection) commitUrl.openConnection();
            connection.setDoOutput(true);
            connection.setRequestMethod("POST");
            connection.setRequestProperty("Content-Type", "application/json");

            JSONObject jsonBody = new JSONObject();
            jsonBody.put("URL", fileUrl);
            jsonBody.put("originalname", originalName);
            jsonBody.put("filemime", mimeType);

            connection.getOutputStream().write(jsonBody.toString().getBytes());

            int responseCode = connection.getResponseCode();
            if (responseCode != HttpURLConnection.HTTP_OK) {
                Log.e(TAG, "Error committing upload, response code: " + responseCode);
            }
        } catch (Exception e) {
            Log.e(TAG, "Error committing upload: " + e.getMessage());
        }
    }
}

File: src/ios/AzureUpload.swift
import Foundation
import UIKit
import AVFoundation
import UserNotifications

@objc(AzureUpload) class AzureUpload: CDVPlugin {

    let STORAGE_URL = "https://arabicschool.blob.core.windows.net/"
    let channelId = "azure_upload_channel"

    @objc(uploadFiles:) 
    func uploadFiles(command: CDVInvokedUrlCommand) {
        let postId = command.arguments[0] as! String
        let sasToken = command.arguments[1] as! String
        let files = command.arguments[2] as! [[String: Any]]
        
        let queue = DispatchQueue(label: "upload.queue", attributes: .concurrent)
        createNotificationChannel()

        for file in files {
            let fileName = file["filename"] as! String
            let originalName = file["originalname"] as! String
            let mimeType = file["mimetype"] as! String
            let binaryData = file["binarydata"] as! String
            let thumbnail = file["thumbnail"] as! String
            
            if mimeType.starts(with: "image") {
                queue.async {
                    self.uploadImage(fileName, binaryData: binaryData, sasToken: sasToken, originalName: originalName, postId: postId)
                }
            } else if mimeType.starts(with: "video") {
                queue.async {
                    self.uploadVideo(fileName, binaryData: binaryData, thumbnail: thumbnail, sasToken: sasToken, originalName: originalName, postId: postId)
                }
            } else {
                queue.async {
                    self.uploadFile(fileName, binaryData: binaryData, sasToken: sasToken, originalName: originalName, postId: postId)
                }
            }
        }
    }

    private func uploadImage(_ fileName: String, binaryData: String, sasToken: String, originalName: String, postId: String) {
        if let imageData = Data(base64Encoded: binaryData) {
            if let image = UIImage(data: imageData) {
                if let pngData = image.pngData() {
                    uploadToAzure(fileName, fileData: pngData, sasToken: sasToken, originalName: originalName, postId: postId)
                }
            }
        }
    }

    private func uploadVideo(_ fileName: String, binaryData: String, thumbnail: String, sasToken: String, originalName: String, postId: String) {
        if let videoData = Data(base64Encoded: binaryData) {
            uploadToAzure(fileName, fileData: videoData, sasToken: sasToken, originalName: originalName, postId: postId)
            
            let tempURL = FileManager.default.temporaryDirectory.appendingPathComponent("tempVideo.mp4")
            do {
                try videoData.write(to: tempURL)
                if let thumbnailData = generateThumbnail(from: tempURL) {
                    uploadToAzure(thumbnail, fileData: thumbnailData, sasToken: sasToken, originalName: originalName, postId: postId)
                }
            } catch {
                print("Error saving video data: \(error.localizedDescription)")
            }
        }
    }

    private func uploadFile(_ fileName: String, binaryData: String, sasToken: String, originalName: String, postId: String) {
        if let fileData = Data(base64Encoded: binaryData) {
            uploadToAzure(fileName, fileData: fileData, sasToken: sasToken, originalName: originalName, postId: postId)
        }
    }

    private func uploadToAzure(_ fileName: String, fileData: Data, sasToken: String, originalName: String, postId: String) {
        let urlString = "\(STORAGE_URL)\(fileName)?\(sasToken)"
        guard let url = URL(string: urlString) else { return }
        
        var request = URLRequest(url: url)
        request.httpMethod = "PUT"
        request.setValue("application/octet-stream", forHTTPHeaderField: "Content-Type")
        
        let task = URLSession.shared.uploadTask(with: request, from: fileData) { _, response, error in
            if let error = error {
                print("Error uploading to Azure: \(error.localizedDescription)")
                self.showNotification(title: "Upload Error", content: "Failed to upload \(originalName)")
            } else {
                self.commitUpload(postId: postId, fileUrl: "\(STORAGE_URL)\(fileName)", originalName: originalName, mimeType: "application/octet-stream")
                self.showNotification(title: "Upload Complete", content: "File \(originalName) uploaded successfully.")
            }
        }
        task.resume()
    }

    private func commitUpload(postId: String, fileUrl: String, originalName: String, mimeType: String) {
        let urlString = "https://personal-fjlz3d21.outsystemscloud.com/uploads/rest/a/commit?postid=\(postId)"
        guard let url = URL(string: urlString) else { return }
        
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let body: [String: Any] = [
            "URL": fileUrl,
            "originalname": originalName,
            "filemime": mimeType
        ]
        
        do {
            let jsonData = try JSONSerialization.data(withJSONObject: body, options: [])
            let task = URLSession.shared.uploadTask(with: request, from: jsonData) { _, response, error in
                if let error = error {
                    print("Error committing upload: \(error.localizedDescription)")
                }
            }
            task.resume()
        } catch {
            print("Error creating JSON for commit: \(error.localizedDescription)")
        }
    }

    private func generateThumbnail(from url: URL) -> Data? {
        let asset = AVAsset(url: url)
        let imageGenerator = AVAssetImageGenerator(asset: asset)
        imageGenerator.appliesPreferredTrackTransform = true
        
        do {
            let thumbnailImage = try imageGenerator.copyCGImage(at: CMTime.zero, actualTime: nil)
            let image = UIImage(cgImage: thumbnailImage)
            return image.pngData()
        } catch {
            print("Error generating thumbnail: \(error.localizedDescription)")
            return nil
        }
    }

    private func createNotificationChannel() {
        if #available(iOS 10.0, *) {
            let center = UNUserNotificationCenter.current()
            center.requestAuthorization(options: [.alert, .sound]) { granted, error in
                if granted {
                    let content = UNMutableNotificationContent()
                    content.title = "Azure Uploads"
                    content.body = "Notifications for file uploads to Azure"
                    content.sound = UNNotificationSound.default
                    let request = UNNotificationRequest(identifier: self.channelId, content: content, trigger: nil)
                    center.add(request) { error in
                        if let error = error {
                            print("Error adding notification request: \(error.localizedDescription)")
                        }
                    }
                }
            }
        }
    }

    private func showNotification(title: String, content: String) {
        if #available(iOS 10.0, *) {
            let content = UNMutableNotificationContent()
            content.title = title
            content.body = content
            content.sound = UNNotificationSound.default
            
            let request = UNNotificationRequest(identifier: UUID().uuidString, content: content, trigger: nil)
            UNUserNotificationCenter.current().add(request) { error in
                if let error = error {
                    print("Error showing notification: \(error.localizedDescription)")
                }
            }
        }
    }
}
